p2p/pex/pex_reactor.go

```diff
@@ -194,6 +194,12 @@ func (r *PEXReactor) AddPeer(p Peer) {
 			return
 		}
 
+		// If the node is reporting a non-routable address then we use the remote IP used in establishing the connection
+		// While this doesn't prevent adding bogus IPs, it does help allow us to grow the address book with routable peers
+		if !addr.Routable() {
```

Comment:
I'm becoming increasingly worried about adding such `if` constructions. In two years, nobody will be able to understand all the intricacies of the code.

Can't we reject the peer if it reports non-routable IP? https://github.com/tendermint/tendermint/blob/5997e75c84f643a65cc411331e1f47201ebb5772/p2p/node_info.go#L122

Reply:
Agree with the sentiment re such conditions but I don't think we can just reject peers for not knowing their own IP (which is quite common on p2p networks, especially for peers behind NATs).

Perhaps what we need to do is fix the address earlier on, like when we first unmarshal the NodeInfo for the peer ? Would it make sense to check if its routable as soon as we receive it, and if not update it to the remote addr? I'm not sure on first take if this will have unintended side effects.

Reply:
> Can't we reject the peer if it reports non-routable IP?

That approach would be undesirable for a few of the reasons @ebuchman has mentioned. Sockets already provide a nice mechanism for the receiver to identify the source without requiring the source to properly self-report its IP address.

I agree that this _fix_ is more of a band-aid and I too would prefer to set this properly earlier when first receiving the NodeInfo. If that is the desired direction I can make the change for review.

Orthogonal to this fix - I would possibly suggest ignoring or removing a node's self-reporting IP address feature. I see this feature as one that is more ripe for miss-use than benefit. But don't want the discussion around that clouding the fix that I believe should happen for nodes reporting non-routable addresses.

Reply:
> Perhaps what we need to do is fix the address earlier on, like when we first unmarshal the NodeInfo for the peer

👍 

> I would possibly suggest ignoring or removing a node's self-reporting IP address feature. 

I'm in favor of removing, but I think we're using it in several places.

Reply:
https://github.com/tendermint/tendermint/blob/master/p2p/transport.go#L469-L506 Would this be the location where the peer node info is first unmarshalled or is there another or multiple locations?

Reply:
This should be the only location. The correct place to add the check is probably https://github.com/tendermint/tendermint/blob/master/p2p/transport.go#L438

Reply:
@melekes I tried to port the change to the referenced area in transport.go; however NodeInfo is not mutable in the required way at that point in the code so I opted for a change in the handshake portion where the correct fields are mutable.

Reply:
> however NodeInfo is not mutable in the required way at that point in the code

it's not mutable, but still can be overwritten

`nodeInfo.ListenAddress =` should work

Reply:
I'd prefer to keep handshake function clean and do not add any non-related functionality.

Reply:
`it's not mutable, but still can be overwritten` I believe I tried that and it doesn't work because the type within the upgrade is `NodeInfo` which does not know about a `ListenAddress` field. Would it be appropriate to cast it?

Reply:
Oh, I see what you mean now.

Reply:
> Would it be appropriate to cast it?

no